main.go

```diff
@@ -0,0 +1,42 @@
+package main
+
+import (
+	"os"
+	"prescript/lib"
+	"strings"
+)
+
+func main() {
+	config, err := lib.GetConfig()
+	if err != nil || config.Subcommand == lib.NoCommand {
+		os.Exit(lib.USER_ERROR)
+	}
+
+	level := "none"
+	if config.Subcommand == lib.PlayCommand {
+		level = strings.ToLower(config.Play.LogLevel)
+	}
+
+	logger, err := lib.NewLogger(level)
+	if err != nil {
+		os.Exit(lib.INTERNAL_ERROR)
+	}
+	defer logger.Close()
+	logger.Debug("successfully parsed arguments and flags")
+	config.Logger = logger
+
+	if config.Subcommand == lib.PlayCommand {
```

Comment:
This feels a little strange to me, not sure why you're using cobra with defined commands for play and record, but passing this value back up to main to determine which logic to run.

Why not have the `Run` function of each Cobra command handle this logic? It could either be a wrapper around some dedicated structs or types. Or you could have methods or functions that meet the Cobra command Run signature.

Reply:
Yeah, I wasn't sure about this. I went with cobra because the end-user experience on the command line is pretty good. However, I wanted to be able to keep the argument/flag processing separate from the rest of the code so I could also test it separately. I think when I started off, I thought there might be certain flags that would be mutually exclusive so I wanted to make sure I didn't introduce any ambiguities.

Only once I got into the weeds with it did I discover the inversion of control of the `Run` function. Not really a huge fan of that because of the aforementioned testing. Perhaps I need to take another look at it.

> It could either be a wrapper around some dedicated structs or types. Or you could have methods or functions that meet the Cobra command Run signature.

I think I understand the 2nd part of this (`main.go` passes in functions to the config method to use for the `play` and `record commands`), but not the first. How would the wrapper dedicated structs/types work?

Reply:
Ah ok, no worries. So, to elaborate on the second point. Cobra has a pretty simple signature for it's `Run` function (https://pkg.go.dev/github.com/spf13/cobra#Command):

```golang
func(cmd *cobra.Command, args []string) {
    // Do Stuff Here
}
```

So you could define your command to reference a method on an existing function that encapsulates some piece of functionality.

As an example:

```golang
func init() {
  rootCmd.AddCommand(playCmd)
}

var playCmd = &cobra.Command{
  Use:   "play [script file] [optional executable]",
  Short: "Runs prescripted responses against an interactive cli",
  Run:  play,
}

func play(cmd *cobra.Command, args []string) {
  // Do Stuff Here
}
```

Or you can do something more advanced and have a structure that embeds a cobra command while also encapsulating your functionality:

```golang
type playCommand struct {
  *cobra.Command

  config cfg.PlayConfig
  run script.Run
  logger utils.Logger
}

func NewPlayCommand(config cfg.PlayConfig, run script.Run, logger utils.Logger) *playCommand {
  p := &playCommand{
    config: config,
    run: run,
    logger: logger,
  }

  p.Command = &cobra.Command{
    Use:   "play [script file] [optional executable]",
    Short: "Runs prescripted responses against an interactive cli",
    Run:  p.runScript,
  }
}

func (p *playCommand) runScript(cmd *cobra.Command, args []string) {
  // Do Stuff Here
}
```

Then in some other file you can create and register the command:

```golang
play := NewPlayCommand(config, script, logger)
rootCmd.AddCommand(play)
```

This is totally untested of course, but I've done similar things with Cobra in the past